File: 2021/day02.py
def part1(commands):

    def forward(n):
        nonlocal hor
        hor += n

    def up(n):
        nonlocal depth
        depth -= n

    def down(n):
        nonlocal depth
        depth += n

    depth, hor = 0, 0
    m = {
        'forward': forward,
        'up': up,
        'down': down,
    }
    for command in commands:
        m[command[0]](int(command[1]))
    print(hor * depth)


def part2(commands):

    def forward(n):
        nonlocal hor, depth
        hor += n
        depth += aim * n

    def up(n):
        nonlocal aim
        aim -= n

    def down(n):
        nonlocal aim
        aim += n

    aim, depth, hor = 0, 0, 0
    m = {
        'forward': forward,
        'up': up,
        'down': down,
    }
    for command in commands:
        m[command[0]](int(command[1]))
    print(hor * depth)


with open("input/day02.txt") as f:
    v = list(map(lambda l: l.split(), f.readlines()))
    part1(v)
    part2(v)

File: 2021/day02.go
package main

import (
	"aoc/utils"
	"fmt"
	"strings"
)

// Part1 of day2
func Part1(cmds []string) {
	depth, hor := 0, 0
	for _, cmd := range cmds {
		splCmd := strings.Split(cmd, " ")
		switch n := utils.ToIntOrDie(splCmd[1]); splCmd[0] {
		case "up":
			depth -= n
		case "down":
			depth += n
		case "forward":
			hor += n
		}
	}
	fmt.Println(depth * hor)
}

// Part2 of day2
func Part2(cmds []string) {
	aim, depth, hor := 0, 0, 0
	for _, cmd := range cmds {
		splCmd := strings.Split(cmd, " ")
		switch n := utils.ToIntOrDie(splCmd[1]); splCmd[0] {
		case "up":
			aim -= n
		case "down":
			aim += n
		case "forward":
			hor += n
			depth += aim * n
		}
	}
	fmt.Println(depth * hor)
}

func main() {
	cmds := utils.ReadFileLinesOrDie("./input/day02.txt")
	Part1(cmds)
	Part2(cmds)
}
